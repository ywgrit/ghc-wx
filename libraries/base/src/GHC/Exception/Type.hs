{-# LANGUAGE Safe #-}
{-# OPTIONS_HADDOCK not-home #-}

-- |
-- Module      :  GHC.Exception.Type
-- Copyright   :  (c) The University of Glasgow, 1998-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Exceptions and exception-handling functions.
--

module GHC.Exception.Type
       ( Exception(..)    -- Class
       , SomeException(..), ArithException(..)
       , divZeroException, overflowException, ratioZeroDenomException
       , underflowException
       ) where

import GHC.Internal.Exception.Type
