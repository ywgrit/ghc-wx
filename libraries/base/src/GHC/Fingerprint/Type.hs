{-# LANGUAGE Safe #-}

-- |
--
-- Module      :  GHC.Fingerprint.Type
-- Copyright   :  (c) The University of Glasgow, 1994-2023
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  ghc-devs@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Fingerprints for recompilation checking and ABI versioning, and
-- implementing fast comparison of Typeable.
--
-- /The API of this module is unstable and not meant to be consumed by the general public./
-- If you absolutely must depend on it, make sure to use a tight upper
-- bound, e.g., @base < 4.X@ rather than @base < 5@, because the interface can
-- change rapidly without much warning.

module GHC.Fingerprint.Type
    (Fingerprint(..)
     ) where

import GHC.Internal.Fingerprint.Type