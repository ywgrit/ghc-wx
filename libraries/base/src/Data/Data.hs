{-# LANGUAGE Safe #-}

-- |
-- Module      :  Data.Data
-- Copyright   :  (c) The University of Glasgow, CWI 2001--2004
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  non-portable (local universal quantification)
--
-- This module provides the 'Data' class with its primitives for
-- generic programming, along with instances for many datatypes. It
-- corresponds to a merge between the previous "Data.Generics.Basics"
-- and almost all of "Data.Generics.Instances". The instances that are
-- not present in this module were moved to the
-- @Data.Generics.Instances@ module in the @syb@ package.
--
-- \"Scrap your boilerplate\" --- Generic programming in Haskell.  See
-- <https://wiki.haskell.org/Research_papers/Generics#Scrap_your_boilerplate.21>.
--

module Data.Data (

        -- * Module Data.Typeable re-exported for convenience
        module Data.Typeable,

        -- * The Data class for processing constructor applications
        Data(
                gfoldl,
                gunfold,
                toConstr,
                dataTypeOf,
                dataCast1,      -- mediate types and unary type constructors
                dataCast2,      -- mediate types and binary type constructors
                -- Generic maps defined in terms of gfoldl
                gmapT,
                gmapQ,
                gmapQl,
                gmapQr,
                gmapQi,
                gmapM,
                gmapMp,
                gmapMo
            ),

        -- * Datatype representations
        DataType,       -- abstract
        -- ** Constructors
        mkDataType,
        mkIntType,
        mkFloatType,
        mkCharType,
        mkNoRepType,
        -- ** Observers
        dataTypeName,
        DataRep(..),
        dataTypeRep,
        -- ** Convenience functions
        repConstr,
        isAlgType,
        dataTypeConstrs,
        indexConstr,
        maxConstrIndex,
        isNorepType,

        -- * Data constructor representations
        Constr,         -- abstract
        ConIndex,       -- alias for Int, start at 1
        Fixity(..),
        -- ** Constructors
        mkConstr,
        mkConstrTag,
        mkIntegralConstr,
        mkRealConstr,
        mkCharConstr,
        -- ** Observers
        constrType,
        ConstrRep(..),
        constrRep,
        constrFields,
        constrFixity,
        -- ** Convenience function: algebraic data types
        constrIndex,
        -- ** From strings to constructors and vice versa: all data types
        showConstr,
        readConstr,

        -- * Convenience functions: take type constructors apart
        tyconUQname,
        tyconModule,

        -- * Generic operations defined in terms of 'gunfold'
        fromConstr,
        fromConstrB,
        fromConstrM

  ) where

import GHC.Internal.Data.Data
import Data.Typeable
