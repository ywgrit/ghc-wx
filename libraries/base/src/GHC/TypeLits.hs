{-# LANGUAGE Safe #-}
{-# LANGUAGE ExplicitNamespaces #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE NoStarIsType #-}

-- |
--
-- GHC's @DataKinds@ language extension lifts data constructors, natural
-- numbers, and strings to the type level. This module provides the
-- primitives needed for working with type-level numbers (the 'Nat' kind),
-- strings (the 'Symbol' kind), and characters (the 'Char' kind). It also defines the 'TypeError' type
-- family, a feature that makes use of type-level strings to support user
-- defined type errors.
--
-- For now, this module is the API for working with type-level literals.
-- However, please note that it is a work in progress and is subject to change.
-- Once the design of the @DataKinds@ feature is more stable, this will be
-- considered only an internal GHC module, and the programmer interface for
-- working with type-level data will be defined in a separate library.
--
-- @since 4.6.0.0
--

module GHC.TypeLits
    (-- *  Kinds
     Natural,
     Nat,
     Symbol,
     -- *  Linking type and value level
     KnownNat(natSing),
     natVal,
     natVal',
     KnownSymbol(symbolSing),
     symbolVal,
     symbolVal',
     KnownChar(charSing),
     charVal,
     charVal',
     SomeNat(..),
     SomeSymbol(..),
     SomeChar(..),
     someNatVal,
     someSymbolVal,
     someCharVal,
     sameNat,
     sameSymbol,
     sameChar,
     decideNat,
     decideSymbol,
     decideChar,
     OrderingI(..),
     cmpNat,
     cmpSymbol,
     cmpChar,
     -- **  Singleton values
     SNat,
     SSymbol,
     SChar,
     pattern SNat,
     pattern SSymbol,
     pattern SChar,
     fromSNat,
     fromSSymbol,
     fromSChar,
     withSomeSNat,
     withSomeSSymbol,
     withSomeSChar,
     withKnownNat,
     withKnownSymbol,
     withKnownChar,
     -- *  Functions on type literals
     type (<=),
     type (<=?),
     type (+),
     type (*),
     type (^),
     type (-),
     type Div,
     type Mod,
     type Log2,
     AppendSymbol,
     CmpNat,
     CmpSymbol,
     CmpChar,
     ConsSymbol,
     UnconsSymbol,
     CharToNat,
     NatToChar,
     -- *  User-defined type errors
     TypeError,
     ErrorMessage(..)
     ) where

import GHC.Internal.TypeLits
