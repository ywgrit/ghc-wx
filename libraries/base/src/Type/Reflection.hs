{-# LANGUAGE Safe #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE MagicHash #-}

-- |
--
-- Module      :  Type.Reflection
-- Copyright   :  (c) The University of Glasgow, CWI 2001--2017
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  non-portable (requires GADTs and compiler support)
--
-- This provides a type-indexed type representation mechanism, similar to that
-- described by,
--
-- * Simon Peyton-Jones, Stephanie Weirich, Richard Eisenberg,
-- Dimitrios Vytiniotis. "<https://www.microsoft.com/en-us/research/wp-content/uploads/2016/08/dynamic.pdf A reflection on types>".
-- /Proc. Philip Wadler's 60th birthday Festschrift/, Edinburgh (April 2016).
--
-- The interface provides 'I.TypeRep', a type representation which can
-- be safely decomposed and composed. See "Data.Dynamic" for an example of this.
--
-- @since 4.10.0.0
--

module Type.Reflection
    (-- *  The Typeable class
     Typeable,
     typeRep,
     withTypeable,
     -- *  Propositional equality
     (:~:)(Refl),
     (:~~:)(HRefl),
     -- *  Type representations
     -- **  Type-Indexed
     TypeRep,
     pattern TypeRep,
     typeOf,
     pattern App,
     pattern Con,
     pattern Con',
     pattern Fun,
     typeRepTyCon,
     rnfTypeRep,
     eqTypeRep,
     decTypeRep,
     typeRepKind,
     splitApps,
     -- **  Quantified
     SomeTypeRep(..),
     someTypeRep,
     someTypeRepTyCon,
     rnfSomeTypeRep,
     -- *  Type constructors
     TyCon,
     tyConPackage,
     tyConModule,
     tyConName,
     rnfTyCon,
     -- *  Module names
     Module,
     moduleName,
     modulePackage,
     rnfModule
     ) where

import GHC.Internal.Type.Reflection
