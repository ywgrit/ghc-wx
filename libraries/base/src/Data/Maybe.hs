{-# LANGUAGE Safe #-}

-- |
--
-- Module      :  Data.Maybe
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  portable
--
-- The Maybe type, and associated operations.
--

module Data.Maybe
    (Maybe(Nothing, Just),
     maybe,
     isJust,
     isNothing,
     fromJust,
     fromMaybe,
     listToMaybe,
     maybeToList,
     catMaybes,
     mapMaybe
     ) where

import GHC.Internal.Data.Maybe