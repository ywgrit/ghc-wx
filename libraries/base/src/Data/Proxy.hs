{-# LANGUAGE Safe #-}

-- |
--
-- Module      :  Data.Proxy
-- License     :  BSD-style (see the LICENSE file in the distribution)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  portable
--
-- Definition of a Proxy type (poly-kinded in GHC)
--
-- @since 4.7.0.0

module Data.Proxy
    (Proxy(..),
     asProxyTypeOf,
     KProxy(..)
     ) where

import GHC.Internal.Data.Proxy