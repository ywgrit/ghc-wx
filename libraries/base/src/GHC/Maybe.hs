{-# OPTIONS_HADDOCK not-home #-}

-- |
-- Module      :  GHC.Maybe
-- Copyright   :  (c) The University of Glasgow 1997-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  ghc-devs@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- The 'Maybe' type.

module GHC.Maybe
    (Maybe(..)) where

import GHC.Internal.Maybe
