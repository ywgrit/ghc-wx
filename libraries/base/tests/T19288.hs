{-# LANGUAGE DataKinds, PolyKinds, TypeOperators, Safe, TypeFamilies #-}

module Main where

import Data.Maybe
import Data.Proxy
import Type.Reflection
import GHC.TypeLits

data Dat (x :: Proxy 1) = MkD1

evil :: Maybe (Nat :~~: Symbol)
evil = eqTypeRep (case (typeRepKind (typeRep :: TypeRep Dat)) of
                          (Fun (App _ x) _) -> typeRepKind x)
                 (typeRep :: TypeRep Symbol)


data family Cast k l r
newtype instance Cast Nat l r = CastNat { runCastNat :: l }
newtype instance Cast Symbol l r = CastSymbol { runCastSymbol :: r }

{-# NOINLINE castHelper #-}
castHelper :: Maybe (a :~~: b) -> Cast a l r -> Cast b l r
castHelper (Just HRefl) = id
castHelper Nothing      = error "No more bug!"

cast :: a -> b
cast = runCastSymbol . castHelper evil . CastNat

main :: IO ()
main = print (cast 'a' :: Int)
