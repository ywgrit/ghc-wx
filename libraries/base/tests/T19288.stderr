T19288: No more bug!
CallStack (from HasCallStack):
  error, called at T19288.hs:25:27 in main:Main
HasCallStack backtrace:
    collectBacktraces, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:92:13 in ghc-internal:GHC.Internal.Exception
    toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:128:3 in ghc-internal:GHC.Internal.Exception



